File: web/src/api/scans.ts
import type { ScanCreateRequest, ScanListResponse, ScanStatusResponse, ScanType } from '../types/scans'

export async function createScanRun(scanType: ScanType, request: ScanCreateRequest): Promise<ScanStatusResponse> {
  const response = await fetch(`/api/scanners/${scanType}/runs`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(request),
  })
  if (!response.ok) {
    const text = await response.text()
    throw new Error(`Failed to create scan: ${response.status} ${text}`)
  }
  return response.json() as Promise<ScanStatusResponse>
}

export async function fetchScanRuns(scanType: ScanType): Promise<ScanListResponse> {
  const response = await fetch(`/api/scanners/${scanType}/runs`)
  if (!response.ok) {
    const text = await response.text()
    throw new Error(`Failed to load scans: ${response.status} ${text}`)
  }
  return response.json() as Promise<ScanListResponse>
}

export async function fetchScanRun(scanType: ScanType, scanId: string): Promise<ScanStatusResponse> {
  const response = await fetch(`/api/scanners/${scanType}/runs/${scanId}`)
  if (!response.ok) {
    const text = await response.text()
    throw new Error(`Failed to load scan: ${response.status} ${text}`)
  }
  return response.json() as Promise<ScanStatusResponse>
}

export async function fetchScanResults(scanType: ScanType, scanId: string): Promise<unknown> {
  const response = await fetch(`/api/scanners/${scanType}/runs/${scanId}/results`)
  if (!response.ok) {
    const text = await response.text()
    throw new Error(`Failed to load results: ${response.status} ${text}`)
  }
  return response.json() as Promise<unknown>
}

export async function fetchScanParams(
  scanType: ScanType,
): Promise<{ defaults: Record<string, unknown>; schema: unknown }> {
  const response = await fetch(`/api/scanners/${scanType}/params`)
  if (!response.ok) {
    const text = await response.text()
    throw new Error(`Failed to load params schema: ${response.status} ${text}`)
  }
  return response.json() as Promise<{ defaults: Record<string, unknown>; schema: unknown }>
}

File: src/app/api/routes/scans.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from pathlib import Path

from starlette.responses import FileResponse

from app.api.deps import ApiDependencies, get_deps
from app.scans.models import ScanCreateRequest, ScanListResponse, ScanStatusResponse, ScanType
from app.scans.params import params_model_for_scan_type

router = APIRouter(prefix="/scanners", tags=["scanners"])


@router.post("/{scan_type}/runs", response_model=ScanStatusResponse)
def create_scan_run(
    scan_type: ScanType,
    request: ScanCreateRequest,
    deps: ApiDependencies = Depends(get_deps),
) -> ScanStatusResponse:
    return deps.scan_jobs.create_scan(scan_type, request)


@router.get("/{scan_type}/runs", response_model=ScanListResponse)
def list_scan_runs(
    scan_type: ScanType,
    deps: ApiDependencies = Depends(get_deps),
) -> ScanListResponse:
    items = deps.scan_jobs.list_scans(scan_type, limit=10)
    return ScanListResponse(items=items)


@router.get("/{scan_type}/runs/{scan_id}", response_model=ScanStatusResponse)
def get_scan_run(
    scan_type: ScanType,
    scan_id: str,
    deps: ApiDependencies = Depends(get_deps),
) -> ScanStatusResponse:
    item = deps.scan_jobs.get_scan(scan_type, scan_id)
    if item is None:
        raise HTTPException(status_code=404, detail="Scan run not found")
    return item


@router.get("/{scan_type}/runs/{scan_id}/results")
def download_scan_results(
    scan_type: ScanType,
    scan_id: str,
    deps: ApiDependencies = Depends(get_deps),
) -> FileResponse:
    item = deps.scan_jobs.get_scan(scan_type, scan_id)
    if item is None:
        raise HTTPException(status_code=404, detail="Scan run not found")
    if not item.results_json_path:
        raise HTTPException(status_code=404, detail="Results not available")
    file_path = Path(item.results_json_path)
    if not file_path.exists():
        raise HTTPException(status_code=404, detail="Results file not found")
    return FileResponse(path=str(file_path), media_type="application/json", filename="results.json")


@router.get("/{scan_type}/params")
def get_scan_params_schema(scan_type: ScanType) -> dict:
    """
    Used by the web UI to render themed parameter widgets instead of raw JSON.
    """
    model = params_model_for_scan_type(scan_type)
    return {
        "scan_type": scan_type,
        "defaults": model().model_dump(),
        "schema": model.model_json_schema(),
    }
